src/actions.rs

```diff
@@ -0,0 +1,497 @@
+    pub fn new(id: &str) -> Self {
+        Self {
+            id: id.to_string(),
+            actions: Vec::new(),
+        }
+    }
+}
+
+impl From<TouchActions> for ActionSequence {
+    fn from(ta: TouchActions) -> Self {
+        ActionSequence(WDActions::ActionSequence {
+            id: ta.id,
+            actions: WDActions::ActionsType::Pointer {
+                parameters: WDActions::PointerActionParameters {
+                    pointer_type: WDActions::PointerType::Touch,
+                },
+                actions: ta.actions.into_iter().map(|x| x.into_item()).collect(),
+            },
+        })
+    }
+}
+
+/// A sequence of actions to be performed.
+///
+/// See the documentation for [`Actions`] for more details.
+#[derive(Debug)]
+pub struct ActionSequence(pub(crate) WDActions::ActionSequence);
+
+/// Each sequence type implements `InputSource` which provides a `pause()` and a `then()`
+/// method. Each call to `pause()` or `then()` represents one tick for this sequence.
+pub trait InputSource: Into<ActionSequence> {
+    /// The action type associated with this `InputSource`.
+    type Action;
+
+    /// Add a pause action to the sequence for this input source.
+    fn pause(self, duration: Duration) -> Self;
+
+    /// Add the specified action to the sequence for this input source.
+    fn then(self, action: Self::Action) -> Self;
+}
+
+impl InputSource for NullActions {
+    type Action = NullAction;
+
+    fn pause(self, duration: Duration) -> Self {
+        self.then(NullAction::Pause { duration })
+    }
+
+    fn then(mut self, action: Self::Action) -> Self {
+        self.actions.push(action);
+        self
+    }
+}
+
+impl InputSource for KeyActions {
+    type Action = KeyAction;
+
+    fn pause(self, duration: Duration) -> Self {
+        self.then(KeyAction::Pause { duration })
+    }
+
+    fn then(mut self, action: Self::Action) -> Self {
+        self.actions.push(action);
+        self
+    }
+}
+
+impl InputSource for MouseActions {
+    type Action = PointerAction;
+
+    fn pause(self, duration: Duration) -> Self {
+        self.then(PointerAction::Pause { duration })
+    }
+
+    fn then(mut self, action: Self::Action) -> Self {
+        self.actions.push(action);
+        self
+    }
+}
+
+impl InputSource for PenActions {
+    type Action = PointerAction;
+
+    fn pause(self, duration: Duration) -> Self {
+        self.then(PointerAction::Pause { duration })
+    }
+
+    fn then(mut self, action: Self::Action) -> Self {
+        self.actions.push(action);
+        self
+    }
+}
+
+impl InputSource for TouchActions {
+    type Action = PointerAction;
+
+    fn pause(self, duration: Duration) -> Self {
+        self.then(PointerAction::Pause { duration })
+    }
+
+    fn then(mut self, action: Self::Action) -> Self {
+        self.actions.push(action);
+        self
+    }
+}
+
+/// A list of action sequences to be performed via `Client::perform_actions()`
```

Comment:
`perform_actions` has been renamed to just `perform`. If you make this
```suggestion
/// A list of action sequences to be performed via [`Client::perform()`]
```
It should be auto-linked by rustdoc, and it'll also tell us if the name ever changes in the future.